src/programs.c

```diff
@@ -988,26 +1104,99 @@ static __always_inline int enter_exec(syscall_pattern_type_t sp, int fd,
 
     bpf_map_update_elem(&telemetry_ids, &pid, &id, BPF_ANY);
 
-    // explicit null check to satisfy the verifier
-    if (!filename)
+Exit:
+    return 0;
+}
+
+static __always_inline ptelemetry_event_t enter_exec_4_8(syscall_pattern_type_t sp, int fd,
+                                      const char __user *filename,
+                                      const char __user *const __user *argv,
+                                      const char __user *const __user *envp,
+                                      int flags, struct pt_regs *ctx, u32 ppid,
+                                      u32 luid, const char __user *exe, u32 len)
+{
+    u64 id = 0; // reuse ID to save space
+    ptelemetry_event_t pev = bpf_map_lookup_elem(&telemetry_stack, (u32 *) &id);
+    if (!pev) // this should never happen, but the verifier complains otherwise
+    {
+
         goto Exit;
+    }
+    // this pointer passing and memcpy is to put the struct on the stack and satisfy the verifier
+    telemetry_event_t sev = {0};
+    __builtin_memcpy(&sev, pev, sizeof(telemetry_event_t));
+    ptelemetry_event_t ev = &sev;
 
+    u64 p_t = bpf_get_current_pid_tgid();
+    u32 t = p_t >> 32;
+    id = (u64) bpf_map_lookup_elem(&telemetry_ids, &t);
+    if (id)
+    {
+        __builtin_memcpy(&id, (void*)id, sizeof(u64));
+        goto Pwd;
+    }
+
+    id = bpf_get_prandom_u32();
     ev->id = id;
     ev->done = FALSE;
-    ev->telemetry_type = TE_EXE_PATH;
+    ev->telemetry_type = 0,
+    ev->u.v.value[0] = '\0';
     ev->u.v.truncated = FALSE;
-    long count = 0;
-    count = bpf_probe_read_str(&ev->u.v.value, VALUE_SIZE, (void *)filename);
-    if (count == VALUE_SIZE)
-    {
-        ev->u.v.truncated = TRUE;
-    }
+
+    FILL_TELEMETRY_SYSCALL_EVENT(ev, sp);
+    u32 pid = pid_tgid >> 32;
+    ev->u.syscall_info.ppid = ppid;
+    ev->u.syscall_info.luid = luid;
     push_telemetry_event(ev);
 
-    bpf_tail_call(ctx, &tail_call_table, 0);
+    bpf_map_update_elem(&telemetry_ids, &pid, &id, BPF_ANY);
 
+    u32 count = 0;
+    char br = 0;
+    //READ_VALUE_N(ev, TE_EXE_PATH, exe, 5); // TODO: is this necessary?
+
+Pwd:;
+    u64 offset = 0;
```

Comment:
Seems like this spot has a lot of repetitive code. Could we declare and inline function or macro to clean this up a bit? The `SET_OFFSET` and `FOLLOW_PTR` macros might work I think if you made them a bit more generic.

Reply:
ea1cf31